### PR TITLE
Add Snowplow Media plugin with APIs to track media events (close #1176)

### LINKED ISSUE #1176: Add Snowplow Media plugin with APIs to track media events
The goal is to provide a plugin with media player-agnostic APIs to track media events. The plugin should give a specification of how to track media events regardless of which player is used. It will make use of universal event and entity schemas for media tracking.

### PR DESCRIPTION
**Pull request by matus-tomlein**
Issue #1176 

Adds a new plugin for tracking media events using `trackMediaXXX()` functions. The APIs are intended to be media player agnostic and usable for all kinds of video players (in contrast with our other plugins that target YouTube, Vimeo, ...).

## Tracking API

Simplified example of how one would use the plugin:

```javascript
newTracker("ns1", "https://...", { plugins: [SnowplowMediaPlugin(),] });

const id = 'media-tracking-id';

startMediaTracking({ // required to call this for the tracker to keep track of the media
  id,
  player: { // properties of the media player context entity
    currentTime: 0,
    duration: 100,
  },
  session: true, // enabled by default, adds a session context entity
  pings: true, // enable ping events
  boundaries: [25, 50, 75], // enable percentage progress events and send them at the specific boundaries
});

trackMediaPlay({ id });
trackMediaPause({ id, media: { currentTime: 100 } });

// ...

endMediaTracking({ id }); // clears the state for the media tracking
```

## Documentation

The documentation is deployed as a preview here:

1. [JavaScript tracker plugin](https://deploy-preview-423--snowplow-docs.netlify.app/docs/collecting-data/collecting-from-own-applications/javascript-trackers/javascript-tracker/javascript-tracker-v3/plugins/media/).
2. [Browser tracker plugin](https://deploy-preview-423--snowplow-docs.netlify.app/docs/collecting-data/collecting-from-own-applications/javascript-trackers/browser-tracker/browser-tracker-v3-reference/plugins/media/).

The [draft PR for the docs is here](https://github.com/snowplow/documentation/pull/423).

## Schemas

The plugin makes use of some existing as well as new schemas:

1. It uses the [`media_player` schema](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/media_player/jsonschema/1-0-0) for a context entity that is added to all events and describes the current state of the media player.
3. There are a few new context entities that are currently [under review in this PR](https://github.com/snowplow/iglu-central/pull/1270):
    * `media_player_session` which holds an ID for all media events from the player and also some statistics about the playback computed on the tracker.
    * `media_player_ad` and `media_player_ad_break` are used to describe the currently played ads.
4. It also uses new schemas for the media player events. Instead of using a single event schema (as our other plugins do), we decided to create separate schemas for each event type. This makes it easier to filter and subscribe to specific events. There is an open PR where the event schemas [are under review here](https://github.com/snowplow/iglu-central/pull/1283).

## Example app

Probably the easiest way to understand how the data looks like and how it is tracked is through an example app that I have built. The app is currently deployed at this address (this is just temporary, once reviewed I'd like to host it on Github pages for the examples repo so that anyone can inspect it to understand how media tracking works): https://snowplow-media-demo.tomlein.org/media

The app shows a media player and a table with a list of tracked events under it. The table is automatically updated as new events are tracked.

I have created a PR on the examples repository to add this app: https://github.com/snowplow-incubator/snowplow-javascript-tracker-examples/pull/56

### CODE REVIEW

File: plugins/browser-plugin-media/src/schemata.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/schemata.ts:12]: I think we can generate a dict at runtime from the enum instead of this switch, to save a bit of space from strings. I did a similar thing in one of the media trackers, though it requires a bit of type finagling. You should be able to do something like:

```
type MediaPlayerEventTypeKeyMap = {
  [key in MediaPlayerEventType]: string;
};

// Allows us to create an empty object of type
let mediaPlayerEventTypeKeyMap = {} as MediaPlayerEventTypeKeyMap;

// Fill the empty object with key/values and add the prefix to the values
const prefix = "media_player_event_";
Object.keys(MediaPlayerEventType).forEach((k) => {
  mediaPlayerEventTypeKeyMap[MediaPlayerEventType[k as keyof typeof MediaPlayerEventType]]
    = prefix + MediaPlayerEventType[k as keyof typeof MediaPlayerEventType]
});


console.log(mediaPlayerEventTypeKeyMap[MediaPlayerEventType.Ready]); // "media_player_event_ready" 
```

File: plugins/browser-plugin-media/src/schemata.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/schemata.ts:3]: `iglu:com.snowplowanalytics.snowplow/media_player` could probably also be placed into a variable rather than repeating the string. (and maybe even `/jsonschema/1-0-0` too?)

File: plugins/browser-plugin-media/src/schemata.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/schemata.ts:11]: Good point, it really was too verbose! I couldn't do the 1:1 translation of the event types to schema names because a couple of types have different names (ad first quartile, midpoint, third quartile). So I left a switch for them, but otherwise just decided to concat strings with the event type.

File: plugins/browser-plugin-media/src/schemata.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/schemata.ts]: Good idea! I have extracted the schema prefix and suffix into a variable and concat it with the schema names.

File: plugins/browser-plugin-media/src/types.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/types.ts:33]:
```suggestion
  /** Media player event fired periodically during main content playback, regardless of other API events that have been sent. */
```

😉

File: plugins/browser-plugin-media/src/mediaTracking.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/mediaTracking.ts:183]: This method seems to perform two fairly distinct functions:
- controlling the state of `this.isSeeking`
- checking if an `eventType` should be tracked

I see why, as we don't want to track a `SeekStart` if we are already seeking, but maybe we could either give the method a name that better captures all the functions it performs, or extract the state-setting into it's own small method to call from within `shouldTrackEvent`, something like:

```
private updateSeekState(eventType: MediaPlayerEventType): void {
    if (eventType == MediaPlayerEventType.SeekStart) {
        if (this.isSeeking) {
            return false;
        }

        this.isSeeking = true;
    } else if (eventType == MediaPlayerEventType.SeekEnd) {
        this.isSeeking = false;
    }
}

private shouldTrackEvent(eventType: MediaPlayerEventType): boolean {
    updateSeekState(eventType);

    if (this.captureEvents !== undefined && !this.isSeeking && !this.captureEvents.includes(eventType)) {
      return false;
    }
}
```

File: plugins/browser-plugin-media/package.json
Comment (greg-el) [inline plugins/browser-plugin-media/package.json:12]: I reckon you deserve to take credit for your work 😉

File: plugins/browser-plugin-media/src/schemata.ts
Comment (greg-el) [reply plugins/browser-plugin-media/src/schemata.ts:11]: Yeah that's much tider than my suggestion too! Perhaps a small comment to explain why the switch is still required, otherwise looks good!

File: plugins/browser-plugin-media/src/schemata.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/schemata.ts:23]:
```suggestion
export function getMediaPlayerEventSchema(eventType: MediaPlayerEventType): string {
```
bit of a nitpick, but it's nice to see at a glance at the signature what/if type a function is expected to return

File: plugins/browser-plugin-media/src/sessionTracking.ts
Comment (greg-el) [inline plugins/browser-plugin-media/src/sessionTracking.ts:23]:
```suggestion
  update(eventType?: MediaPlayerEventType, mediaPlayer: MediaPlayer, adBreak?: MediaPlayerAdBreak) {
```

Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/sessionTracking.ts:23]: That would be better, but after merging the suggestion I had to take it back, because TypeScript complained that the first property can't be made optional.

**Review by jethron — COMMENTED** [review]
Gonna have to trust your testing on the state machine aspects in sessionStats and adTracking. :sweat_smile:

Otherwise mostly looks really good, exciting plugin!

I don't have a lot of experience with the other media plugins to know if this will integrate well, so LGTM if LGT other reviewers.

File: plugins/browser-plugin-media/test/api.test.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/test/api.test.ts]: Good point, changed it.

File: plugins/browser-plugin-media/package.json
Comment (matus-tomlein) [reply plugins/browser-plugin-media/package.json]: I don't like naming individuals either, went for `Snowplow Analytics Ltd (https://snowplow.io/)`.

File: plugins/browser-plugin-media/package.json
Comment (matus-tomlein) [reply plugins/browser-plugin-media/package.json]: Good catch! Removed lodash from dependencies.

File: plugins/browser-plugin-media/src/api.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/api.ts]: Fixed!

File: plugins/browser-plugin-media/src/core.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/core.ts]: Good point, have changed the check for null. I don't think we need to deal with recursively checking the properties yet, currently there are no such schemas in media tracking, so probably too early to do it now.

File: plugins/browser-plugin-media/src/core.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/core.ts]: Fixed!

File: plugins/browser-plugin-media/src/core.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/core.ts]: Good point! Have changed it (needed to make the interfaces extend `Record<string, unknown>` but that should be fine).

File: plugins/browser-plugin-media/src/schemata.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/schemata.ts]: Fixed!

File: plugins/browser-plugin-media/src/mediaTracking.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/mediaTracking.ts]: Fixed!

File: plugins/browser-plugin-media/src/mediaTracking.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/mediaTracking.ts]: Fixed!

File: plugins/browser-plugin-media/src/api.ts
Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/api.ts:109]: That's a nice trick! There are two things that are making a bit nervous about changing that though:

1. We changed some of the events (such as playback rate changed, quality changed) to also contain properties in the event body. So these need to have their own function and can't be bound to the `trackMediaEvent` function.
2. The `args` argument of the `trackMediaEvent` accepts all kinds of properties including `MediaTrackAdArguments` which we don't want to allow in all of the track methods – you should be able to pass an ad entity in `trackMediaPlay`.

File: plugins/browser-plugin-media/package.json
Comment (matus-tomlein) [reply plugins/browser-plugin-media/package.json]: I think Paul was the only one who had the access so we might not be able to get hold of it anymore... Anyway, it seems better to me to avoid using bitly, so I changed it to the Github repository of the tracker. Would be nice to do it also across all the plugins and trackers in the repo.

File: plugins/browser-plugin-media/src/api.ts
Comment (jethron) [reply plugins/browser-plugin-media/src/api.ts:109]: Ah fair enough, this is possible with bind but ends up being uglier than the explicit functions anyway.

e.g.
```typescript
export const trackMediaBufferEnd = trackMediaEvent.bind<null, MediaPlayerEventType | undefined, [MediaTrackArguments & CommonMediaEventProperties, string[]?], void>(null, MediaPlayerEventType.BufferEnd);
```

So fine as is, I think.

File: plugins/browser-plugin-media/src/api.ts
Comment (jethron) [reply plugins/browser-plugin-media/src/api.ts:70]: I think this is OK. Can we promote the 30s to a constant and document/comment that updating it is a breaking change for downstream models? If it is silently updated and models assumed 30s in it's absence we'll suddenly break analysis downstream so just want that documented clearly.

Comment (matus-tomlein) [reply plugins/browser-plugin-media/src/api.ts:71]: That makes sense, I have [documented it internally here](https://github.com/snowplow/snowplow-javascript-tracker/blob/issue/1176-media_plugin/plugins/browser-plugin-media/src/pingInterval.ts#L3-L8).

**Review by greg-el — APPROVED** [review]
LGTM! Integrating this with the Vimeo plugin was smooth sailing, great job!

**Review by jethron — APPROVED** [review]
LGTM!

**Pull request merged**
